db/hints/sync_point_service.cc

```diff
@@ -0,0 +1,442 @@
+                // If abort_source was engaged, this will throw - that's what we want
+                co_await sleep_abortable(follower_status_check_interval, combined_as);
+            }
+        } catch (...) {
+            // TODO: Print error?
+            service_logger.debug("Task ({} | coordinator): got an exception: {}", id, std::current_exception());
+            if (!combined_as.abort_requested()) {
+                combined_as.request_abort();
+            }
+            throw;
+        }
+
+        service_logger.debug("Task ({} | coordinator): the node {} has successfully replayed its hints", id, follower);
+    });
+}
+
+future<> sync_point_service::wait_for_local_hints(
+        abort_source& as,
+        sync_point_id id,
+        std::vector<gms::inet_address> target_nodes) {
+    
+    assert(this_shard_id() == 0);
+
+    // If an exception occurs on one shard, we are no longer interested
+    // in waiting for other shards, so we need a sharded abort source.
+    // Gate is needed because abort_all_shards spawns discarded futures
+    // which needs to be waited on.
+    struct abort_source_and_gate {
+        seastar::gate gate;
+        abort_source as;
+    };
+    sharded<abort_source_and_gate> sharded_abort_source;
+    co_await sharded_abort_source.start();
+
+    auto abort_all_shards = [&sharded_abort_source, id] {
+        auto& local = sharded_abort_source.local();
+        if (!local.gate.try_enter()) {
+            return;
+        }
+
+        // Waited on by the shard-local gate
+        (void)sharded_abort_source.invoke_on(0, [&sharded_abort_source, id] (auto& local0) {
+            // To prevent O(smp::count^2) futures from being generated,
+            // first check if we already aborted on shard 0, and if not,
+            // abort on all shards
+            if (local0.as.abort_requested()) {
+                service_logger.debug("Task ({} | follower): abort requested from above", id);
+                return make_ready_future<>();
+            }
+            local0.as.request_abort();
+            return sharded_abort_source.invoke_on_others([] (auto& local) {
+                if (!local.as.abort_requested()) {
+                    local.as.request_abort();
+                }
+            });
+        }).finally([&local] {
+            local.gate.leave();
+        });
+    };
+
+    auto sub = as.subscribe([&abort_all_shards] () noexcept {
+        abort_all_shards();
+    });
+
+    co_await container().invoke_on_all([&sharded_abort_source, &abort_all_shards, id, &target_nodes] (sync_point_service& sps) {
+        sps.check_if_is_running();
+
+        return parallel_for_each(sps._managers, [&sharded_abort_source, &target_nodes, id] (manager& mgr) {
+            service_logger.debug("Task ({} | follower | shard {}): going to wait on shard for the following nodes: {}", this_shard_id(), id, target_nodes);
+            return mgr.wait_until_hints_are_replayed(sharded_abort_source.local().as, target_nodes);
```

Comment:
"hints are replayed" is something that might never be true, if more hints are generated. Maybe we need to wait until all hints that were written before we started waiting are replayed.

Reply:
> Maybe we need to wait until all hints that were written before we started waiting are replayed.

This is the current semantic. I can change the name to something more precise (e.g. `wait_until_current_hints_are_replayed`?)

Reply:
Changed the name to `wait_until_current_hints_are_replayed`